src/Serilog.Sinks.MSSqlServer/Sinks/MSSqlServer/Platform/SqlInsertBatchWriter.cs

```diff
@@ -0,0 +1,86 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Data;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Serilog.Debugging;
+using Serilog.Events;
+using Serilog.Sinks.MSSqlServer.Output;
+using static System.FormattableString;
+
+namespace Serilog.Sinks.MSSqlServer.Platform
+{
+    internal class SqlInsertBatchWriter : ISqlBulkBatchWriter
+    {
+        private readonly string _tableName;
+        private readonly string _schemaName;
+        private readonly ISqlConnectionFactory _sqlConnectionFactory;
+        private readonly ILogEventDataGenerator _logEventDataGenerator;
+
+        public SqlInsertBatchWriter(
+            string tableName,
+            string schemaName,
+            ISqlConnectionFactory sqlConnectionFactory,
+            ILogEventDataGenerator logEventDataGenerator)
+        {
+            _tableName = tableName ?? throw new ArgumentNullException(nameof(tableName));
+            _schemaName = schemaName ?? throw new ArgumentNullException(nameof(schemaName));
+            _sqlConnectionFactory = sqlConnectionFactory ?? throw new ArgumentNullException(nameof(sqlConnectionFactory));
+            _logEventDataGenerator = logEventDataGenerator ?? throw new ArgumentNullException(nameof(logEventDataGenerator));
+        }
+
+        public async Task WriteBatch(IEnumerable<LogEvent> events, DataTable dataTable)
```

Comment:
It's a bit leaky that `ISqlBulkBatchWriter` provides a `DataTable` here and I assume still builds up all the columns in it, but probably not easy to change that.

Reply:
def looks a bit leaky, but probably something I'd rather not tackle as an outsider unless required.